engine/access/rpc/backend/backend_accounts.go

```diff
@@ -126,6 +255,142 @@ func (b *backendAccounts) getAccountAtBlock(
 	}
 }
 
+func (b *backendAccounts) getAccountBalanceAtBlock(
+	ctx context.Context,
+	address flow.Address,
+	blockID flow.Identifier,
+	height uint64,
+) (uint64, error) {
+	switch b.scriptExecMode {
+	case IndexQueryModeExecutionNodesOnly:
+		account, err := b.getAccountFromAnyExeNode(ctx, address, blockID)
+		if err != nil {
+			b.log.Debug().Err(err).Msgf("failed to get account balance at blockID: %v", blockID)
+			return 0, err
+		}
+		return account.Balance, nil
+
+	case IndexQueryModeLocalOnly:
+		accountBalance, err := b.scriptExecutor.GetAccountBalance(ctx, address, height)
+		if err != nil {
+			b.log.Debug().Err(err).Msgf("failed to get account balance at blockID: %v", blockID)
+			return 0, err
+		}
+
+		return accountBalance, nil
+	case IndexQueryModeFailover:
+		localAccountBalance, localErr := b.scriptExecutor.GetAccountBalance(ctx, address, height)
+		if localErr == nil {
+			return localAccountBalance, nil
+		}
+		execResult, execErr := b.getAccountFromAnyExeNode(ctx, address, blockID)
+		if execErr != nil {
+			return 0, execErr
+		}
+
+		return execResult.Balance, nil
+
+	default:
+		return 0, status.Errorf(codes.Internal, "unknown execution mode: %v", b.scriptExecMode)
+	}
+}
+
+func (b *backendAccounts) getAccountKeysAtBlock(
+	ctx context.Context,
+	address flow.Address,
+	blockID flow.Identifier,
+	height uint64,
+) ([]flow.AccountPublicKey, error) {
+	switch b.scriptExecMode {
+	case IndexQueryModeExecutionNodesOnly:
+		account, err := b.getAccountFromAnyExeNode(ctx, address, blockID)
+		if err != nil {
+			b.log.Debug().Err(err).Msgf("failed to get account keys at blockID: %v", blockID)
+			return nil, err
+		}
+		return account.Keys, nil
+	case IndexQueryModeLocalOnly:
+		accountKeys, err := b.scriptExecutor.GetAccountKeys(ctx, address, height)
+		if err != nil {
+			b.log.Debug().Err(err).Msgf("failed to get account keys at height: %d", height)
+			return nil, err
+		}
+
+		return accountKeys, nil
+	case IndexQueryModeFailover:
+		localAccountKeys, localErr := b.scriptExecutor.GetAccountKeys(ctx, address, height)
+		if localErr == nil {
+			return localAccountKeys, nil
+		}
+
+		account, err := b.getAccountFromAnyExeNode(ctx, address, blockID)
+		if err != nil {
+			return nil, err
+		}
+
+		return account.Keys, nil
+
+	default:
+		return nil, status.Errorf(codes.Internal, "unknown execution mode: %v", b.scriptExecMode)
+	}
+
+}
+
+func (b *backendAccounts) getAccountKeyAtBlock(
+	ctx context.Context,
+	address flow.Address,
+	blockID flow.Identifier,
+	keyIndex uint64,
+	height uint64,
```

Comment:
consider this to keep the `blockID` and `height` consistent as the last arguments
```suggestion
	ctx context.Context,
	address flow.Address,
	keyIndex uint64,
	blockID flow.Identifier,
	height uint64,
```